feat!: machine api

nderjung opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [x] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This PR represents a major refactor to KraftKit's interface for manipulating VM instances.  At its core, a new top-level API package (available at `kraftkit.sh/api`) exposes externally accessible KraftKit internal APIs via the introduction of a new framework, [The Zip API Object Framework](https://api.zip).  The framework allows the definition of APIs in a Kubernetes-like fashion without requiring code generation.  Therefore it is possible to generate traditional CRD-like objects (objects usually defined with a "Spec" and "Status") in a simple way.  The interface is defined simply by introducing a relevant Spec-and-Status structs and using the Zip API generic:

```go
import zip "api.zip"

type (
	// Machine is the mutable API object that represents a machine instance.
	Machine = zip.Object[MachineSpec, MachineStatus]

	// MachineList is the mutable API object that represents a list of machine
	// instances.
	MachineList = zip.ObjectList[MachineSpec, MachineStatus]
)
```

The use of the Zip API Object Framework additionally provides a configurable client which is protocol agnostic, allowing the interface to the object to be defined by the implementer.  One potential client implementation could simply be an inline function call, a remote HTTP request, gRPC call, etc.  Unlike the traditional Kubernetes client, these calls can be inline code function calls.  This allows KraftKit to use a controller-pattern without having to make external calls.  It also paves the way for a future implementation to use the same API with minimal configuration change to invoke the call remotely, if KraftKit is ever daemonized. For example, for the `Create` method of the `MachineService`, we wrap the implementing method with the Zip API client:

```go
create, err := zip.NewMethodClient(ctx, impl.Create, opts...)
if err != nil {
	return nil, err
}
```

The Zip API client is versatile and allows for "before" and "after" wrapping of calls which are defined by the implementer through the [`WithStore` option](https://github.com/api-zip/api-zip/blob/c0ff264a51ca77abf5daa2329019b03c64cf09b9/client_options.go#L47-L145).  This allows us to save the state of API calls through the new embedded store, allowing for content rehydration and local state of the API.  Before, the state was always tailored to the machine. Now, when future APIs are defined, we no longer have to think about its management. This refactor migrates the older store to a new generic embedded store compatible with the Zip API Framework.

This PR introduces the both the definition of the interface for manipulating VMs, or simply "machines" by KraftKit, which essentially defines the configuration used within the context of the current machine driver, as well as necessary implementation methods which manipulate the API in a meaningful way:

```go
// MachineService is the interface of available methods which can be performed
// by an implementing machine platform driver.
type MachineService interface {
	Create(context.Context, *Machine) (*Machine, error)
	Start(context.Context, *Machine) (*Machine, error)
	Pause(context.Context, *Machine) (*Machine, error)
	Stop(context.Context, *Machine) (*Machine, error)
	Update(context.Context, *Machine) (*Machine, error)
	Delete(context.Context, *Machine) (*Machine, error)
	Get(context.Context, *Machine) (*Machine, error)
	List(context.Context, *MachineList) (*MachineList, error)
	Watch(context.Context, *Machine) (chan *Machine, chan error, error)
	Logs(context.Context, *Machine) (chan string, chan error, error)
}
```

This large refactor deprecates the `machine/driver` package in favour of the new top-level API package interface and the implementing interfaces.  To make this switch, the QEMU driver implementation has been adjusted to match the service implementation defined in `api/machine/v1alpha1`.  As a result, all uses of driver in subcommands are adjusted simultaneously to keep the project in working condition. 

Part of this refactor adjusts how the `kraft run` command works due to the nature of the change of the API. Particularly, the split between "kernel arguments" and "application arguments" is made more prevelant through the API.  As such, the args supplied to `kraft run` has changed, which now has an additional `--kernel-args` flag.  Additionally, this refactor introduces a new package, `machine/platform`, aimed at a). naming supported underlying systems using a similar naming convention to Unikraft (already present via `unikraft/plat`) and b). to replace the machine driver model.  Since the goal is to make the choice between the runtime platform highly configurable (and hosts can support different platforms), a mechanism for instantiating and providing choice between platform implementations is provided via "strategy" concept. This strategization of the platform allows for quickly switching between different implementations at a high-level.

Finally, miscellaneous additional fixes are included, including ones which affect `kraft rm` and `kraft stop` such that they are UNIX-like and mimic the output of equivalent commands from `docker`. 

GitHub-Depends: https://github.com/unikraft/kraftkit/pull/503
GitHub-Depends: https://github.com/unikraft/kraftkit/pull/505

nderjung committed:
fix(kraft): Use internal waitgroup package

Remove duplicate code by using the already present waitgroup packge.
This makes the implementation identicatl to its use in
kraft/events/events.go

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(qemu): Squash the license and copyright notice preamble

This commit "squashes" the license and copyright preamble in files
associated with this sign off.  The preamble already includes the SPDX
license identifier which essentially duplicates the text in question.
Additionally, this license, in accordance with the license, is
distributed with the source code via the `LICENSE.md` file which is
accompanied in this repository.

Additionally, copyright notice is provided to The KraftKit Authors,
which entitles new developers who contribute commits to the project an
associated copyright with regard to their contributions without having
to specifically update the license and copyright preamble.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(qemu): Prefix qemu-system-* structures/constants

This commit prefixes all files related to direct interfaces with the
qemu-system-* binaries (and therefore offer the Go abstractions which
map to the CLI flags) with qemu_.  This ensures that files related to
this part of the package vs. other parts (such as configuration,
machine driver, etc.) are clearly separated.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(api): Introduce top-level API package

This commit introduces a top-level API package which exposes KraftKit
internal APIs via the Zip API Object Framework[0].  The framework allows
the definition of APIs in a Kubernetes-like fashion without requiring
code generation.  Therefore it is possible to generate traditional
CRD-like objects (objects usually defined with a "Spec" and "Status") in
a simple way.

The use of the Zip API Object Framework additionally provides a
configurable client which is protocol agnostic, allowing the interface
to the object to be defined by the implementer.  One potential client
implementation could simply be an inline function call, a remote HTTP
request, gRPC call, etc.

This initial commit introduces the definition of the "Machine Interface"
for KraftKit, which essentially defines the configuration used within
the context of the current machine driver.  Whilst this API is not used
within KraftKit as of this commit, this provides a base towards how the
driver interface can exist in the future: independent, well-defined,
versioned.

[0]: https://api.zip

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(store): Add a Zip API-compatible store implementation

This store is intended to be used internally to KraftKit as an
embeddable location for objects.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(name): Introduce machine name generator with custom names

This commit introduces a new internal name generator based on Docker's
name generator[0].  For the purpose of being context-specific to the
project, and for a little fun, the package has been adjusted to include
famous gorillas, monkeys, chimpanzee, orangutans, bonobos and baboons
whose names were identified by lightly parsing Wikipedia.

[0]: https://github.com/moby/moby/pkg/namesgenerator

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(platform): Introduce machine platform strategy

This commit introduces a new package, `machine/platform`, aimed at a).
naming supported underlying systems using a similar naming convention to
Unikraft (already present via `unikraft/plat`) and b). to eventually
replace the machine driver model.  Since the goal is to make the choice
between the runtime platform highly configurable (and hosts can support
different platforms), a mechanism for instantiating and providing choice
between platform implementations is provided via "strategy" concept.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(qemu): Migrate init method to its own file

Since the init method contains largely boilerplate for future gob
registrations, the method is quite large.  Migrate to its own file to
keep the driver file smaller and more focused.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(qemu): Register `QemuConfig` with `gob`

This structure will eventually be saved to the internal store and so
requires registation as a binary object.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(set): Introduce additional set helper methods

This commit creates a distinction between finding any or exactly 1 item
in a set.  Before, the use of the string set `Contains` would return if
any items were found.  This may have unintended effect when we are
expecting to find exact match.

Necessary adjustments of the new methods are made in the project.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor!: Use the API-defined machine interface

This large refactor deprecates the `machine/driver` package in favour of
the new top-level API package.  To make this switch, the qemu driver
implementation has been adjusted to match the service implementation
defined in `api/machine/v1alpha1`.

All uses of driver in subcommands are adjusted simultaneously to keep
the project in working condition.

Part of this refactor adjusts how the `kraft run` command works due to
the nature of the change of the API. Particularly, the split between
"kernel arguments" and "application arguments" is made more prevalent
through the API.  As such, the args supplied to `kraft run` has changed,
which now has an additional `--kernel-args` flag.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(qemu): Rename the service implementing file

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on machine/platform/detect.go:
```diff
@@ -0,0 +1,145 @@
+	return ret, nil
+}
+
+// Detect returns the hypervisor and system mode in the context to the deremined
```
`determined`

craciunoiuc commented on machine/qemu/qemu.go:
```diff
@@ -675,52 +608,51 @@ func (qd *QemuDriver) State(ctx context.Context, mid machine.MachineID) (state m
 		activeProcess, err = process.IsRunning()
 		if err != nil {
-			state = machine.MachineStateDead
+			fmt.Printf("1111\n")
```
`2222`

craciunoiuc commented on machine/qemu/qemu.go:
```diff
@@ -675,52 +608,51 @@ func (qd *QemuDriver) State(ctx context.Context, mid machine.MachineID) (state m
-			if err = qd.dopts.Store.SaveMachineState(mid, state); err != nil {
-				return
-			}
+			fmt.Printf("seting state\n")
```
👀

nderjung committed:
fix(run): Set cwd earlier so it can be used for full paths

Use the invocation of `os.Getwd()` earlier so that it can be used to set
the fully path of the kernel.  Additionally, only set the kernel name in
the API spec reference for the kernel.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(run): Remove additional log entries to make usage more UNIX-like

This commit returns the machine name such that the output of `kraft run`
can be pushed through a UNIX pipe, allowing for greater ease-of-use.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(run): Stop the machine before removing it

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(run): Request a context shutdown after receiving an event error

The loop was previously broken, but the context was never cancelled,
which can result in malformed method shutdowns which rely on listening
to `ctx.Done()`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(run): Use the package initrd if set and no CLI option was supplied

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(ps): Remove "driver" from table

Driver is the same as platform and therefore a redundant column.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(qemu): Fix out-of-bounds for maximum discarded logs

This commit fixes an issue where the discarded part of the log is
greater or equal to the length of the size of the log.  Simply perform a
check before performing the subslice.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(qemu): Correct use of Go json tags for `QemuMachine`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(qemu): Do not append with `WithAppend`

This confuses the use of the method option, like all other instances, it
should directly set the value rather than manipulate the existing value.
This can lead to misuse.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

nderjung committed:
fix(stop): Fix pointer mismatch and adjust output to be more UNIX-like

This commit addresses the `kraft stop` command which would incorrectly
stop instances.  Additionally, upon success, the name of the machine is
returned which aligns with `docker stop` and also alows the command to
be used in a UNIX pipe.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(rm): Stop instance before remove and adjust to make more UNIX-like

This commit checks whether the instance is online and stops it before
removing the supplied instance.  Additionally, to make the command more
UNIX-like and similar to `docker rm`, upon success, the name of the
instance is returned such that it can be used in a UNIX pipe.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(stop): Handle the use of no-positional-arguments use of `kraft run`

The command expects either 1-to-many machine names or IDs or the use of
the `--all` flag.  Handle appropriately.

Signed-off-by: Alexander Jung <alex@unikraft.io>
